### PR TITLE
[inductor] fix index.Tensor fallback

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144736

The original issue is we see accuracy problem in a meta internal model [meta internal link](https://fb.workplace.com/groups/1075192433118967/posts/1567334737238065/).  The debugging is hard but the root cause is relatively simple. The root cause is that the model has mix-device inputs for index.Tensor which causes Inductor to fallback. And the meta kernel for index.Tensor returns a tensor with inconsistent strides to the eager kernel.

The following code snippet
```
import torch
from torch._subclasses import FakeTensorMode

device = "cuda"

x = torch.randn((24, 16, 32, 32), device=device).to(memory_format=torch.channels_last)
x = x.view(2, 12, 16, 32, 32)

i1 = torch.arange(2).unsqueeze(-1)
i2 = torch.argsort(torch.rand(2, 12), dim=-1)[:, :3]

print(f"Eager stride: {x[i1, i2].stride()}")

mode = FakeTensorMode()
with mode:
    f_x = mode.from_tensor(x)
    f_i1 = mode.from_tensor(i1)
    f_i2 = mode.from_tensor(i2)
    f_out = f_x[f_i1, f_i2]
    print(f"Meta stride: {f_out.stride()}")
```

would output:
```
Eager stride: (49152, 16384, 1, 512, 16)
Meta stride: (49152, 16384, 1024, 32, 1)
```

In this PR, I fix the problem to run eager kernel to get the index.Tensor fallback's output layout. A better solution would be to change meta/eager kernel implementation so that their output layout matches. But I'm not sure how to properly do that.
In the index.Tensor meta kernel, we always produce dense output:  https://github.com/pytorch/pytorch/blob/6d56277682715e56cfdfcaff6f770acebda966d7/torch/_meta_registrations.py#L3184 . While the eager kernel seems to leverage TensorIteratorBase to decide some dimension permutation: https://github.com/pytorch/pytorch/blob/6d56277682715e56cfdfcaff6f770acebda966d7/aten/src/ATen/TensorIterator.cpp#L232-L308 .  We can duplicate this logic to the meta kernel implementation if we really want meta matches eager. I can follow up on this if people have strong opinion to do this.

And here is an issue https://github.com/pytorch/pytorch/issues/144717 for asserting size/strides for fallback kernels. With that, the issue debugged here would be much easier to root cause.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @BoyuanFeng

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -24,6 +24,8 @@
 from typing import Tuple
 from unittest.mock import patch
 
+import einops
```
Comment (jansel) [inline test/inductor/test_torchinductor.py:27]: Should handle einops not being installed?  Perhaps import locally in the test?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -6682,7 +6682,10 @@ def tensor_to_layout(output: torch.Tensor):  # type: ignore[no-untyped-def]
 
     @classmethod
     def create(cls, kernel, *args, **kwargs):  # type: ignore[no-untyped-def]
-        fake_incorrect_kernels = (aten._fused_moving_avg_obs_fq_helper_functional,)
+        fake_incorrect_kernels = (
```
Comment (eellison) [inline torch/_inductor/ir.py:6685]: This doesn't work with symints, would it be possible to fix the meta instead ?

**Review by eellison — COMMENTED** [review]
might regress dynamic shapes

File: torch/_inductor/ir.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/ir.py:6685]: To fix the meta, we should duplicate [this C++ function]( https://github.com/pytorch/pytorch/blob/6d56277682715e56cfdfcaff6f770acebda966d7/aten/src/ATen/TensorIterator.cpp#L232-L308) into torch/_meta_implementations.py right?

Comment (ezyang) [reply torch/_inductor/ir.py:6685]: I feel we should already have reorder_dimensions but I do not see it. Check if `compute_elementwise_output_logical_to_physical_perm` is what you want.

**Review by jansel — APPROVED** [review]
